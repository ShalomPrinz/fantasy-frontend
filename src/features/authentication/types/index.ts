export { parseUser } from "./User";
export type { LoginUser, RegisterUser } from "./User";
