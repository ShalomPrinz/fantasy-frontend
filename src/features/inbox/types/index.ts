export type { Message } from "./Inbox";
