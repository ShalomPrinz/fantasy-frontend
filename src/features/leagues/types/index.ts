export { isLeagueAdmin, parseLeague } from "./League";
export type { CreateLeague, DetailedLeague, LeagueInfo } from "./League";
