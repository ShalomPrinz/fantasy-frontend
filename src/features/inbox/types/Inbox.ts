import { LeagueInfo } from "../../leagues";

interface Message {
  id: string;
  from: string;
  league: LeagueInfo;
}

export type { Message };
