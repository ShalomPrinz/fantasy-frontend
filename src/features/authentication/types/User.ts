import { User } from "../../../types";
import { Team } from "../../team";

export interface RegisterUser {
  fullName: string;
  username: string;
  email: string;
  password: string;
}

export interface LoginUser {
  email: string;
  password: string;
}

/** Receive user from response, Returns app user object */
export function parseUser(user: any) {
  const { id, inbox, leagues, username, team } = user;
  const userTeam = new Team(team, true);
  const appUser: User = {
    id,
    inbox,
    leagues: leagues || [],
    name: username,
    team: userTeam,
  };
  return appUser;
}
