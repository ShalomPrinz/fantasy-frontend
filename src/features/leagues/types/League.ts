import { Team } from "../../../types";

interface Member {
  id: string;
  name: string;
  team: Team;
  role: string;
}

interface LeagueInfo {
  id: string;
  membersCount: number;
  name: string;
}

interface DetailedLeague {
  id: string;
  members: Member[];
  name: string;
}

interface CreateLeague {
  name: string;
}

function isLeagueAdmin(league: DetailedLeague | undefined, userId: string) {
  const member = league?.members.find((m) => m.id === userId);
  return member?.role === "admin";
}

/** Receive member from response, Returns app member object */
function parseMember(member: any) {
  const { id, username, team, role } = member;
  const memberTeam = new Team(team, true);
  const appMember: Member = {
    id,
    name: username,
    team: memberTeam,
    role,
  };
  return appMember;
}

/** Receive league from response, Returns app league object */
export function parseLeague(league: any) {
  if (typeof league === "undefined") return undefined;
  const { id, name, members } = league;
  const leagueMembers = members.map((m: any) => parseMember(m));
  const appLeague: DetailedLeague = {
    id,
    name,
    members: leagueMembers,
  };
  return appLeague;
}

export { isLeagueAdmin };
export type { CreateLeague, DetailedLeague, LeagueInfo };
