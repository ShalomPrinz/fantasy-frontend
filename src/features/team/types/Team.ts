import { isPlayerRole } from "../../../types";
import type { Player, PlayerRole } from "../../../types";

type Players = {
  [key in PlayerRole]: Player[];
};

class Team {
  readonly players: Players = {
    GK: [],
    DEF: [],
    MID: [],
    ATT: [],
  };
  count: number;
  isUserTeam: boolean;

  constructor(players: Player[], isUserTeam?: boolean) {
    this.count = 0;
    this.isUserTeam = Boolean(isUserTeam);
    players?.forEach((p) => this.addPlayer(p));
  }

  addPlayer(p: Player) {
    if (!isPlayerRole(p.role) || this.contains(p)) return false;

    this.players[p.role].push(p);
    this.count += 1;
    return true;
  }

  removePlayer(p: Player) {
    if (!isPlayerRole(p.role) || !this.contains(p)) return false;

    const index = this.players[p.role].findIndex((pl) => pl.id === p.id);
    this.players[p.role].splice(index, 1);
    this.count -= 1;
    return true;
  }

  contains({ id, role }: Player) {
    if (!isPlayerRole(role)) return false;
    return this.players[role].findIndex((p) => p.id === id) !== -1;
  }

  clone() {
    const cloned = new Team([], this.isUserTeam);
    Object.values(this.players).forEach((arr) =>
      arr.forEach((p) => cloned.addPlayer(p))
    );
    return cloned;
  }
}

export default Team;
