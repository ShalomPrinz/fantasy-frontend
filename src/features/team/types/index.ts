import Team from "./Team";

export { Team };
export type { PlayerPairs } from "./Team";
