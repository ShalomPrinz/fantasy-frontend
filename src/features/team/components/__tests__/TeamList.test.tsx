import { mockTeam, render } from "setupTests";

import { TeamList } from "../";
import { Team } from "../../types";
import { missingIdPlayer, missingTeamPlayer, player } from "./players";

describe("TeamList", () => {
  it("should render TeamList component", () => {
    const team = new Team([player]);
    mockTeam(team);

    const { asFragment } = render(<TeamList />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("Runtime Errors", () => {
    it("should render with damaged data: no player team", () => {
      //@ts-expect-error should miss team property
      const team = new Team([missingTeamPlayer]);
      mockTeam(team);

      const { asFragment } = render(<TeamList />);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should render with damaged data: no player id", () => {
      //@ts-expect-error should miss id property
      const team = new Team([missingIdPlayer]);
      mockTeam(team);

      const { asFragment } = render(<TeamList />);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should cause uncaught error and render Error Boundary", () => {
      //@ts-expect-error useTeamState is undefined and TeamList calls it as a function
      mockTeam(undefined);

      const { asFragment } = render(<TeamList />);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
