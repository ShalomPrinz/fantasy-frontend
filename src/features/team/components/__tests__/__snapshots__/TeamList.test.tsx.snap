// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamList Runtime Errors should cause uncaught error and render Error Boundary 1`] = `
<DocumentFragment>
  <div
    class="text-center my-3 p-4 bg-danger text-white w-75 mx-auto"
  >
    <h2>
      Team List Not Available
    </h2>
    <button
      class="fs-4 bg-white rounded"
      type="button"
    >
      Refresh
    </button>
  </div>
</DocumentFragment>
`;

exports[`TeamList Runtime Errors should render with damaged data: no player id 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           GK 
        </h1>
        <div
          class="d-flex align-items-center ps-5"
        >
          <img
            class=""
            src="Barcelona.png"
            style="width: 60px;"
            title="Barcelona"
          />
          <h4
            class="text-truncate ps-2 my-0"
          >
             Ter Stegen 
          </h4>
        </div>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           DEF 
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           MID 
        </h1>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           ATT 
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TeamList Runtime Errors should render with damaged data: no player team 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           GK 
        </h1>
        <div
          class="d-flex align-items-center ps-5"
        >
          <img
            class=""
            src="Error.png"
            style="width: 60px;"
          />
          <h4
            class="text-truncate ps-2 my-0"
          >
             Ter Stegen 
          </h4>
        </div>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           DEF 
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           MID 
        </h1>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           ATT 
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TeamList should render TeamList component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           GK 
        </h1>
        <div
          class="d-flex align-items-center ps-5"
        >
          <img
            class=""
            src="Barcelona.png"
            style="width: 60px;"
            title="Barcelona"
          />
          <h4
            class="text-truncate ps-2 my-0"
          >
             Ter Stegen 
          </h4>
        </div>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           DEF 
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           MID 
        </h1>
      </div>
      <div
        class="w-50 col"
      >
        <h1
          class="centered-flex pt-4"
        >
           ATT 
        </h1>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
